src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp

```diff
@@ -0,0 +1,355 @@
+        // y = mx + b where x = rt or mz, m = slope, b = left intensity
+        // sign of delta_int will determine line direction
+        // background += delta_int / delta_pos * (it->getPos() - left) + int_l;
+        UInt n_points = 0;
+        for (auto it=p.PosBegin(left); it!=p.PosEnd(right); ++it, ++n_points)
+        {
+          background += it->getPos();
+        }
+        background = (background - n_points * p.PosBegin(left)->getPos()) * delta_int / delta_pos + n_points * int_l;
+      }
+    }
+    else if (baseline_type_ == "vertical_division")
+    {
+      if (integration_type_ == "trapezoid" || integration_type_ == "simpson")
+      {
+        background = delta_pos * std::min(int_r, int_l);
+      }
+      else if (integration_type_ == "intensity_sum")
+      {
+        UInt n_points = 0;
+        for (auto it=p.PosBegin(left); it!=p.PosEnd(right); ++it, ++n_points)
+        { }
+        background = std::min(int_r, int_l) * n_points;
+      }
+    }
+    background_area_ = background;
+  }
+
+  template <typename PeakContainerT>
+  void PeakIntegrator::integratePeak_(PeakContainerT p, const double& left, const double& right)
+  {
+    peak_area_ = 0.0;
+    peak_height_ = -1.0;
+    peak_apex_pos_ = -1.0;
+    UInt n_points = 0;
+    for (auto it=p.PosBegin(left); it!=p.PosEnd(right); ++it, ++n_points)
+    {
+      if (peak_height_ < it->getIntensity())
+      {
+        peak_height_ = it->getIntensity();
+        peak_apex_pos_ = it->getPos();
+      }
+    }
+
+    if (integration_type_ == "trapezoid")
+    {
+      for (auto it=p.PosBegin(left); it!=p.PosEnd(right)-1; ++it)
+      {
+        peak_area_ += ((it+1)->getPos() - it->getPos()) * ((it->getIntensity() + (it+1)->getIntensity()) / 2.0);
+      }
+    }
+    else if (integration_type_ == "simpson")
+    {
+      if (n_points < 3)
+      {
+        LOG_DEBUG << std::endl << "Error in integratePeak: number of points must be >=3 for Simpson's rule" << std::endl;
+        return;
+      }
+      if (n_points % 2)
+      {
+        peak_area_ = simpson(p.PosBegin(left), p.PosEnd(right));
+      }
+      else
+      {
+        double areas[4] = {-1.0, -1.0, -1.0, -1.0};
+        areas[0] = simpson(p.PosBegin(left), p.PosEnd(right) - 1);   // without last point
+        areas[1] = simpson(p.PosBegin(left) + 1, p.PosEnd(right));   // without first point
+        if (p.begin() <= p.PosBegin(left) - 1)
+        {
+          areas[2] = simpson(p.PosBegin(left) - 1, p.PosEnd(right)); // with one more point on the left
+        }
+        if (p.PosEnd(right) < p.end())
+        {
+          areas[3] = simpson(p.PosBegin(left), p.PosEnd(right) + 1); // with one more point on the right
+        }
+        UInt valids = 0;
+        for (auto area : areas)
+        {
+          if (area != -1.0)
+          {
+            peak_area_ += area;
+            ++valids;
+          }
+        }
+        peak_area_ /= valids;
+      }
+    }
+    else
```

Comment:
I suggest
```
else if (integration_type_ == "intensity_sum")
{
}
else
{
  throw ...
}
```
to ensure that we have 3 distinct ways of integrating and throw an exception if any other way is used. I think otherwise it will be hard to add additional integration techniques as it will not be clear whether they are explicitly supported if we always have a "fallback option" that is used if no other method fits.